iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+
+/// <summary>
+/// Respond to a writable property request received from the service.
+/// 
+/// Property under the root component:
+/// "reported": {
+///     "targetTemperature": {
+///         "value": 21.3,
+///         "ac": 200,
+///         "av": 3,
+///         "ad": "complete"
+///     }
+/// }
+/// 
+/// Property under a specified component:
+/// "reported": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": {
+///       "value": 23,
+///       "ac": 200,
+///       "av": 3,
+///       "ad": "complete"
+///     }
+///   }
+/// }
+/// </summary>
+/// <param name="propertyName">The name of the twin property to update.</param>
+/// <param name="propertyValue">The value of the twin property to update.</param>
+/// <param name="componentName">The name of the component in which the property is defined. Can be null for property defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, WritableProperty propertyValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Subscribe to writable property event for all properties under any component.
+///
+/// Payload received for the root component:
+/// "desired" :
+/// {
+///   "targetTemperature" : 21.3,
+///   "targetHumidity" : 80
+/// },
+/// "$version" : 3
+///
+/// Payload received for a specific component:
+/// "desired": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": 21.3,
+///     "targetHumidity": 80
+///   }
+/// },
+/// "$version" : 3
+/// 
+/// </summary>
+/// <param name="propertyActionAsTwinCollection">The action to be taken when a writeable property event is received.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public void SubscribeToWritablePropertyEvent(Action<TwinCollection> propertyActionAsTwinCollection, CancellationToken cancellationToken = default);
+```
+
+<details>
+<summary>WritableProperty</summary>
+
+```csharp
+/// <summary>
+/// Empty constructor.
+/// </summary>
+public WritablePropertyResponse() { }
+
+/// <summary>
+/// Convenience constructor for specifying the properties.
+/// </summary>
+/// <param name="propertyValue">The unserialized property value.</param>
+/// <param name="ackCode">The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.</param>
+/// <param name="ackVersion">The acknowledgement version, as supplied in the property update request.</param>
+/// <param name="ackDescription">The acknowledgement description, an optional, human-readable message about the result of the property update.</param>
+public WritablePropertyResponse(object propertyValue, int ackCode, long ackVersion, string ackDescription = null)
```

Comment:
Why `object` here for the param, but `dynamic` in the property?

Why this over generic types?

Reply:
Have we tested this with various types to ensure it serializes the way we want?

Reply:
This is a copy-paste issue; I copied over the properties and the constructor from different files and they used different types.  I expect this to be an object.

I've not tested this out yet, no. That is the next task.

Reply:
Originally this was a generic type, with a constraint to use some fancy serializer implementation that wasn't great. Moving to dynamic (or object) gave a very simplistic design that meant you could implement one property subscription and get multiple values. 

However, **I agree that generic types would work well for this** and we could have customers use both strongly typed classes and dynamic interchangably. The only caveat here is we'd need to make sure we have a guard around a situation where 

```c#

// Handle allthethings.gif
SubscribeToWritablePropertyEvent<dynamic>("component",  (componentName,  writableProperty) =>
{
    // Fancy stuff here
});

// Use our concrete class to access properties
SubscribeToWritablePropertyEvent<MyCustomObjectClass>("component", "stronglyTypedProperty", (componentName, propertyName, writableProperty) =>
{
    // Fancy stuff here
});

// What happens here? Override the previous, run both?
SubscribeToWritablePropertyEvent<MyCustomObjectClassThatIsNotRelated>("component", "stronglyTypedProperty", (componentName, propertyName, writableProperty) =>
{
    // Fancy stuff here
});
```

Reply:
Have y'all gotten feedback from Krzysztof from the Azure SDK team about your API design choices? He's historically given us feedback against using dynamic.

I don't prefer object because it can't be deserialized for user-defined types, and requires casting.

I think you two should try creating and testing these API designs before deciding on a proposed interface.